File: supporter-product-data/src/main/scala/com/gu/lambdas/FetchResultsLambda.scala
package com.gu.lambdas

import com.amazonaws.services.lambda.runtime.Context
import com.gu.lambdas.FetchResultsLambda.fetchResults
import com.gu.model.StageConstructors
import com.gu.model.states.{FetchResultsState, AddSupporterRatePlanItemToQueueState}
import com.gu.model.zuora.response.JobStatus.Completed
import com.gu.okhttp.RequestRunners.configurableFutureRunner
import com.gu.services.{ConfigService, S3Service, ZuoraQuerierService}
import com.gu.supporterdata.model.Stage
import com.typesafe.scalalogging.StrictLogging

import java.time.{ZoneOffset, ZonedDateTime}
import java.time.format.DateTimeFormatter
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.duration.DurationInt

class FetchResultsLambda extends Handler[FetchResultsState, AddSupporterRatePlanItemToQueueState] {
  override protected def handlerFuture(input: FetchResultsState, context: Context) =
    fetchResults(StageConstructors.fromEnvironment, input.jobId, input.attemptedQueryTime)
}

object FetchResultsLambda extends StrictLogging {
  val stage = StageConstructors.fromEnvironment
  val config = ConfigService(stage).load
  val service = new ZuoraQuerierService(config, configurableFutureRunner(60.seconds))

  def fetchResults(stage: Stage, jobId: String, attemptedQueryTime: ZonedDateTime) = {
    logger.info(s"Attempting to fetch results for jobId $jobId")
    for {
      result <- service.getResults(jobId)
      _ = assert(result.status == Completed, s"Job with id $jobId is still in status ${result.status}")
      batch = getValueOrThrow(
        result.batches.headOption,
        s"No batches were returned in the batch query response for jobId $jobId",
      )
      fileId = getValueOrThrow(batch.fileId, s"Batch.fileId was missing in jobId $jobId")
      filename = s"select-active-rate-plans-${attemptedQueryTime.withZoneSameInstant(ZoneOffset.UTC).format(DateTimeFormatter.ISO_LOCAL_DATE_TIME)}.csv"
      fileResponse <- service.getResultFileResponse(fileId)
      _ = assert(
        fileResponse.isSuccessful,
        s"File download for job with id $jobId failed with http code ${fileResponse.code}",
      )
      val contentLength = if (fileResponse.body.contentLength >= 0) Some(fileResponse.body.contentLength) else None
      _ = S3Service.streamToS3(stage, filename, fileResponse.body.byteStream, contentLength)
    } yield {
      logger.info(s"Successfully wrote file $filename to S3 with ${batch.recordCount} records for jobId $jobId")
      if (batch.recordCount == 0)
        ConfigService(stage).putLastSuccessfulQueryTime(attemptedQueryTime)

      AddSupporterRatePlanItemToQueueState(
        filename,
        batch.recordCount,
        processedCount = 0,
        attemptedQueryTime,
      )
    }
  }

  def getValueOrThrow[T](maybeValue: Option[T], errorMessage: String) =
    maybeValue match {
      case Some(value) => value
      case None => throw new RuntimeException(errorMessage)
    }
}

File: supporter-product-data-typescript/src/typescript/lambdas/fetchResultsLambda.ts
import type { Handler } from "aws-lambda";
import { stageFromEnvironment } from "../model/stage";
import type { BatchQueryResponse } from "../model/zuora";
import { ConfigService } from "../services/configService";
import { S3Service } from "../services/s3Service";
import { ZuoraQuerierService } from "../services/zuoraQuerierService";
import type {
  AddSupporterRatePlanItemToQueueState,
  FetchResultsState,
} from "./types";

interface FetchResultsDeps {
  getResults: (jobId: string) => Promise<BatchQueryResponse>;
  getResultFileResponse: (fileId: string) => Promise<Response>;
  uploadToS3: (
    stage: ReturnType<typeof stageFromEnvironment>,
    filename: string,
    body: Uint8Array,
    length: number
  ) => Promise<void>;
  putLastSuccessfulQueryTime: (time: string) => Promise<void>;
}

const toIsoLocalDateTimeUtc = (isoString: string): string => {
  const date = new Date(isoString);
  if (Number.isNaN(date.getTime())) {
    throw new Error(`Invalid attemptedQueryTime: ${isoString}`);
  }
  return date.toISOString().replace("Z", "");
};

const getValueOrThrow = <T>(value: T | undefined, errorMessage: string): T => {
  if (value === undefined) {
    throw new Error(errorMessage);
  }
  return value;
};

const buildDefaultDeps = async (): Promise<FetchResultsDeps> => {
  const stage = stageFromEnvironment();
  const configService = new ConfigService(stage);
  const config = await configService.loadZuoraConfig();
  const zuoraService = new ZuoraQuerierService(config);
  const s3Service = new S3Service();

  return {
    getResults: (jobId) => zuoraService.getResults(jobId),
    getResultFileResponse: (fileId) =>
      zuoraService.getResultFileResponse(fileId),
    uploadToS3: (uploadStage, filename, body, length) =>
      s3Service.streamToS3(uploadStage, filename, body, length),
    putLastSuccessfulQueryTime: (time) =>
      configService.putLastSuccessfulQueryTime(time),
  };
};

export const fetchResults = async (
  event: FetchResultsState,
  deps?: FetchResultsDeps
): Promise<AddSupporterRatePlanItemToQueueState> => {
  const stage = stageFromEnvironment();
  const resolvedDeps = deps ?? (await buildDefaultDeps());

  console.info("Attempting to fetch results", { jobId: event.jobId, stage });

  const result = await resolvedDeps.getResults(event.jobId);
  if (result.status !== "completed") {
    throw new Error(
      `Job with id ${event.jobId} is still in status ${result.status}`
    );
  }

  const batch = getValueOrThrow(
    result.batches[0],
    `No batches were returned in the batch query response for jobId ${event.jobId}`
  );
  const fileId = getValueOrThrow(
    batch.fileId,
    `Batch.fileId was missing in jobId ${event.jobId}`
  );

  const filename = `select-active-rate-plans-${toIsoLocalDateTimeUtc(
    event.attemptedQueryTime
  )}.csv`;

  const fileResponse = await resolvedDeps.getResultFileResponse(fileId);
  if (!fileResponse.ok) {
    throw new Error(
      `File download for job with id ${event.jobId} failed with http code ${fileResponse.status}`
    );
  }

  const contentLength = Number(fileResponse.headers.get("content-length") ?? 0);
  if (contentLength <= 0) {
    throw new Error(
      `Content length of the file for job with id ${event.jobId} is not > 0`
    );
  }

  const fileBytes = new Uint8Array(await fileResponse.arrayBuffer());
  await resolvedDeps.uploadToS3(stage, filename, fileBytes, contentLength);

  if (batch.recordCount === 0) {
    await resolvedDeps.putLastSuccessfulQueryTime(event.attemptedQueryTime);
  }

  console.info("Successfully wrote file to S3", {
    filename,
    recordCount: batch.recordCount,
    jobId: event.jobId,
  });

  return {
    filename,
    recordCount: batch.recordCount,
    processedCount: 0,
    attemptedQueryTime: event.attemptedQueryTime,
  };
};

export const handler: Handler<
  FetchResultsState,
  AddSupporterRatePlanItemToQueueState
> = async (event) => fetchResults(event);
